File: src/tui/boot.js
import { TITLE_ART, colorize, emphasize, fit } from "./theme.js";

const SPINNER = ["|", "/", "-", "\\"];

function statusColor(status) {
  switch (status) {
    case "ready":
      return "brightGreen";
    case "loading":
      return "yellow";
    case "error":
      return "red";
    default:
      return "gray";
  }
}

function meter(status, tick, color) {
  if (status === "ready") {
    return colorize("■■■■", "brightGreen", color);
  }
  if (status === "loading") {
    const fill = (tick % 4) + 1;
    return `${colorize("■".repeat(fill), "green", color)}${colorize("□".repeat(4 - fill), "brightBlack", color)}`;
  }
  if (status === "error") {
    return colorize("■■■■", "red", color);
  }
  return colorize("□□□□", "brightBlack", color);
}

export function renderBootFrame(options = {}) {
  const width = Math.max(72, Math.min(Number(options.width ?? 100), 120));
  const tick = Number(options.tick ?? 0);
  const color = Boolean(options.color);
  const activeStepId = options.activeStepId ?? null;
  const steps = options.steps ?? [];

  const lines = [
    ...TITLE_ART.map((line) => colorize(line, "brightGreen", color)),
    emphasize(colorize("VORKER CONTROL PLANE // VORKER-2 supervisor mesh", "green", color), color),
    colorize("Arrow-led operator shell. Booting agent lanes and replaying the supervisor bus.", "gray", color),
    "─".repeat(Math.max(40, width)),
  ];

  for (const step of steps) {
    const liveStatus = step.id === activeStepId ? "loading" : step.status ?? "pending";
    const spinner = liveStatus === "loading" ? ` ${SPINNER[tick % SPINNER.length]}` : "";
    const statusLabel = colorize(`${liveStatus.toUpperCase()}${spinner}`, statusColor(liveStatus), color);
    const line = `${meter(liveStatus, tick, color)} ${step.label.padEnd(14)} ${step.detail} ${statusLabel}`;
    lines.push(fit(line, width));
  }

  return lines.join("\n");
}

File: crates/vorker-tui/src/boot.rs
use crate::theme::{TITLE_ART, colorize, emphasize, fit};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BootStep {
    pub id: String,
    pub label: String,
    pub status: String,
    pub detail: String,
}

impl BootStep {
    #[must_use]
    pub fn new(id: &str, label: &str, status: &str, detail: &str) -> Self {
        Self {
            id: id.to_string(),
            label: label.to_string(),
            status: status.to_string(),
            detail: detail.to_string(),
        }
    }
}

const SPINNER: [&str; 4] = ["|", "/", "-", "\\"];
const BOOT_EXTRA_LINGER_TICKS: usize = 4;

#[must_use]
pub fn boot_minimum_ticks() -> usize {
    TITLE_ART
        .len()
        .saturating_sub(1)
        .saturating_add(BOOT_EXTRA_LINGER_TICKS)
}

fn meter(status: &str, tick: usize, color: bool) -> String {
    match status {
        "ready" => colorize("[####]", "brightGreen", color),
        "loading" => {
            let fill = (tick % 4) + 1;
            format!("[{}{}]", "#".repeat(fill), ".".repeat(4 - fill))
        }
        "error" => colorize("[!!!!]", "red", color),
        _ => colorize("[....]", "brightBlack", color),
    }
}

pub fn render_boot_frame(
    width: usize,
    tick: usize,
    active_step_id: Option<&str>,
    steps: &[BootStep],
    color: bool,
) -> String {
    let width = width.clamp(72, 120);
    let revealed_lines = (tick + 1).min(TITLE_ART.len());
    let mut lines = TITLE_ART
        .iter()
        .enumerate()
        .map(|(index, line)| {
            if index < revealed_lines {
                colorize(line, "brightGreen", color)
            } else {
                " ".repeat(line.chars().count())
            }
        })
        .collect::<Vec<_>>();
    lines.push(emphasize(
        &colorize("VORKER SHELL // Copilot ACP startup", "green", color),
        color,
    ));
    lines.push(colorize(
        "Loading workspace context, model inventory, and command surface.",
        "gray",
        color,
    ));
    lines.push("-".repeat(width.max(40)));

    for step in steps {
        let live_status = if active_step_id == Some(step.id.as_str()) {
            "loading"
        } else {
            step.status.as_str()
        };
        let spinner = if live_status == "loading" {
            format!(" {}", SPINNER[tick % SPINNER.len()])
        } else {
            String::new()
        };
        let status_label = format!("{}{}", live_status.to_uppercase(), spinner);
        let line = format!(
            "{} {:<14} {} {}",
            meter(live_status, tick, color),
            step.label,
            step.detail,
            status_label
        );
        lines.push(fit(&line, width));
    }

    lines.join("\n")
}
